### PR TITLE
Updated Map load/save to Terraria 1.3 file format

### PR DESCRIPTION
**Pull request by speeder**
Also added some new stuff to "Analysis" but not everything is there.

UI needs to be updated (I DID wrote an update, but managed to delete it by accident)

### CODE REVIEW

**Comment by speeder** [comment]
Note: Zooming on maps with new tiles and walls crash the editor, and saving files with new tiles and walls end in a mangled save file.

So until the xml is updated, I recommend against doing the things mentioned.

File: TEditXna/Terraria/WorldAnalysis.cs [hunk 1]
```diff
@@ -260,7 +267,7 @@ private static void WriteFlags(StreamWriter sb, World world)
             sb.WriteProperty("world.AnglerQuest", world.AnglerQuest);
 
             if (world.UnknownData != null && world.UnknownData.Length > 0)
-                sb.WriteProperty("world.UnknownData", BitConverter.ToString(world.UnknownData));
+                sb.WriteProperty("world.UnknownData", BitConverter.ToString(world.UnknownData));
```
Comment (Walkman100) [inline TEditXna/Terraria/WorldAnalysis.cs]: Extra whitespace...

**Comment by SBlogic** [comment]
Just a note on this:  It's loading a 1.3.0.2 world fine (version ID 147), but complains on my 1.3.0.3 world (version ID 149).  The world still seems to load correctly (haven't tried saving).

**Comment by speeder** [comment]
More info please? I generated a new 1.3.0.3 world but it loaded fine...

What was the message?

**Comment by SBlogic** [comment]
Apologies, I grabbed it and forgot to paste it in.  It was this:

> There was an error reading the world file, do you wish to force it to load anyway?
> 
> WARNING: This may have unexpected results including corrupt world files and program crashes.
> The error is :
> 
> System.IO.FileFormatException: Unexpected Position: Invalid Unknown Section
> 
>    at TEditXNA.Terraria.World.LoadV2(BinaryReader b, String filename, World w) in c:\Users\xxxxxx\Documents\Terraria\speeder\Terraria-Map-Editor-master\Terraria-Map-Editor-master\TEditXna\Terraria\World.FileV2.cs:line 645
>
>    at TEditXNA.Terraria.World.LoadWorld(String filename) in c:\Users\xxxxxx\Documents\Terraria\speeder\Terraria-Map-Editor-master\Terraria-Map-Editor-master\TEditXna\Terraria\World.cs:line 110
> 
> Yes   No   

I grabbed a copy of your fork at around 10AM EDT this morning, but I'll refresh and try it again to see, just in case I missed something.  (Is your fork on GitHub synced with  your local copy?  I missed that a couple days ago myself when I erred in trying to open a new PR for additional changes to settings.xml.)

**Comment by speeder** [comment]
Nope, that message comes from updated code...

Mind giving me the file you tried to open? This error is curious (the "unknown" section is a new section that I have no idea what it do, for now I am assuming it is always 4 bytes, but if your file had one that is bigger or smaller than 4 bytes it would give an error...)

**Comment by speeder** [comment]
I still need your file to confirm, but I think I found out what it is... Seemly Terraria 1.3 supports a new type of tile, that is tile that is sort-of-alive, this is used by the Training Dummy.

If my guess is correct, I wonder how to handle that...

**Comment by SBlogic** [comment]
You can grab my world [here](https://drive.google.com/file/d/0B2W8xKoIkz1xR3ZFWHRGdVRUNG8/view?usp=sharing).  I'll check on the Training Dummy's XML properties.  I'm pretty sure I added that, but I'll verify.

**Comment by speeder** [comment]
thedummy

Indeed, it IS the dummy fault... The last section that until then was "unknown" (thus why the error said the section was unkown) save dummies.

The same section also applies to Item Frame

Item Frame is that thing: http://terraria.gamepedia.com/Item_Frame

The data it saves is the type (0 is dummy, 1 is Item Frame), the Id and the position X and Y (the one I put in the image).

**Comment by SBlogic** [comment]
Ah, so is that just something that needs to be accounted for on the code side, or is there anything I could add to the XML to fix that?  And should I re-up the world with some item frames for testing?

**Comment by speeder** [comment]
Alright, I fixed from my side. It was more trouble than I expected, there are lots of completely weird and unkown data, and I ended having to decompile part of Terraria.exe to figure it out.

Basically the new section stores dummies and item frame, it stores the data I mentioned before (type, id, position), and then by detecting the type, the behaviour changes: If it is a dummy, it stores a number that seemly is an NPC id, the dummy to be actually "hit-able" generates a invisible NPC, this information stores that.

As for the Item Frame, it stores an Item ID, Prefix and Stack Size, so not only it can display the correct item, but you can get the item back.

Now the editor itself will have to be reworked a bit, but I don't think I have the time to meddle with that, it requires some recoding of some more complex innards.

It CAN right now load and save files almost 100% correctly (it seemly is changing one of your bosses from killed to not killed or something, weird O.o, but on most files I tested the loaded and saved file are identical), but I expect Terraria will break if someone use TEdit to edit dummies and item frames.

**Comment by SBlogic** [comment]
Good to know.  I'll refresh my copy of your fork and have a look at it.

I can understand item frames having additional data attached (similar case with Weapon Racks and Mannequins I'd imagine) but It seems odd to me that the Target Dummies would have any need for extra data.  I've never really spent much time trying to figure out how tile data is stored, so that's a bit of a mystery to me.

**Comment by enigmaquip** [comment]
The TargetDummy returning an npc value could be futureproofing, maybe it'll be possible to set the npc you want to attack so you can play with weapons and armor sets and make the best choices..... The Item Frame I would expect would need an new interface similar to the chest tab..... in that it stores an item and a prefix

**Comment by enigmaquip** [comment]
Working on the World Properties Display right now if noone else is. Reorganizing bosses downed
worldprop

**Comment by SBlogic** [comment]
I was planning to dig into updating World Properties after dealing with the XML, but it looks like you've got a great start already.  I'll keep plugging away at the data.  Carry on, @enigmaquip!

**Comment by jgibson02** [comment]
I would've DM'd SBlogic directly but couldn't find a way how, but is there anything somehow who has no real programming knowledge can do to help with the TEdit update?

**Comment by speeder** [comment]
@enigmaquip whoa, your edit is much better than the one I started (I just started to stuff all bosses in the same place, it was looking horrendous).

I like it :)

**Comment by TheHoodedTeddy13** [comment]
Is there any way to get the unfinished copy so I can use it before the milestone?  I need to edit out all of the deserts in my world because the new antlions freak me out.

**Comment by rocket4wheel** [comment]
Here's something pretty bad with the newest commit, It kept on crashing with this error: 
"07-08-2015 14:15: Starting TEdit 3.5.15189.1410 
07-08-2015 14:15: OS: Microsoft Windows NT 6.2.9200.0 
07-08-2015 14:15: .Net >= 4.0 
07-08-2015 14:15: XNA 4.0 
07-08-2015 14:15: Terraria Data Path: C:\Program Files (x86)\Steam\steamapps\common\Terraria\Content 
07-08-2015 14:15: Error - Name cannot begin with the '<' character, hexadecimal value 0x3C. Line 6637, position 5.
   at System.Xml.XmlTextReaderImpl.Throw(Exception e)
   at System.Xml.XmlTextReaderImpl.Throw(String res, String[] args)
   at System.Xml.XmlTextReaderImpl.ParseAttributes()
   at System.Xml.XmlTextReaderImpl.ParseElement()
   at System.Xml.XmlTextReaderImpl.ParseElementContent()
   at System.Xml.XmlTextReaderImpl.Read()
   at System.Xml.Linq.XContainer.ReadContentFrom(XmlReader r)
   at System.Xml.Linq.XContainer.ReadContentFrom(XmlReader r, LoadOptions o)
   at System.Xml.Linq.XElement.ReadElementFrom(XmlReader r, LoadOptions o)
   at System.Xml.Linq.XElement.Load(XmlReader reader, LoadOptions options)
   at System.Xml.Linq.XElement.Load(String uri, LoadOptions options)
   at TEditXNA.Terraria.World.LoadObjectDbXml(String file) in c:\Users\Matthew\Desktop\Terraria-Map-Editor-7b05b247164e54d794b8405599cab10d05995c19\TEditXna\Terraria\World.Settings.cs:line 122
   at TEditXNA.Terraria.World..cctor() in c:\Users\Matthew\Desktop\Terraria-Map-Editor-7b05b247164e54d794b8405599cab10d05995c19\TEditXna\Terraria\World.Settings.cs:line 41 
07-08-2015 14:15: Error - The type initializer for 'TEditXNA.Terraria.World' threw an exception.
   at TEditXNA.Terraria.World.get_TileBricks() 
07-08-2015 14:15: Error - Exception has been thrown by the target of an invocation.
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.RuntimePropertyInfo.GetValue(Object obj, BindingFlags invokeAttr, Binder binder, Object[] index, CultureInfo culture)
   at System.Reflection.RuntimePropertyInfo.GetValue(Object obj, Object[] index)
   at System.Windows.Markup.StaticExtension.GetFieldOrPropertyValue(Type type, String name, Object& value)
   at System.Windows.Markup.StaticExtension.ProvideValue(IServiceProvider serviceProvider)
   at MS.Internal.Xaml.Runtime.ClrObjectRuntime.CallProvideValue(MarkupExtension me, IServiceProvider serviceProvider) 
07-08-2015 14:15: Error - Provide value on 'System.Windows.Markup.StaticExtension' threw an exception.
   at System.Windows.Markup.WpfXamlLoader.Load(XamlReader xamlReader, IXamlObjectWriterFactory writerFactory, Boolean skipJournaledProperties, Object rootObject, XamlObjectWriterSettings settings, Uri baseUri)
   at System.Windows.Markup.WpfXamlLoader.LoadBaml(XamlReader xamlReader, Boolean skipJournaledProperties, Object rootObject, XamlAccessLevel accessLevel, Uri baseUri)
   at System.Windows.Markup.XamlReader.LoadBaml(Stream stream, ParserContext parserContext, Object parent, Boolean closeStream)
   at System.Windows.Application.LoadBamlStreamWithSyncInfo(Stream stream, ParserContext pc)
   at System.Windows.Application.LoadComponent(Uri resourceLocator, Boolean bSkipJournaledProperties)
   at System.Windows.Application.DoStartup()
   at System.Windows.Application.<.ctor>b__1(Object unused)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at MS.Internal.Threading.ExceptionFilterHelper.TryCatchWhen(Object source, Delegate method, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at MS.Internal.Threading.ExceptionFilterHelper.TryCatchWhen(Object source, Delegate method, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   at MS.Win32.UnsafeNativeMethods.DispatchMessage(MSG& msg)
   at System.Windows.Threading.Dispatcher.PushFrameImpl(DispatcherFrame frame)
   at System.Windows.Threading.Dispatcher.PushFrame(DispatcherFrame frame)
   at System.Windows.Threading.Dispatcher.Run()
   at System.Windows.Application.RunDispatcher(Object ignore)
   at System.Windows.Application.RunInternal(Window window)
   at System.Windows.Application.Run(Window window)
   at TEditXna.App.Main() in c:\Users\Matthew\Desktop\Terraria-Map-Editor-7b05b247164e54d794b8405599cab10d05995c19\TEditXna\obj\x86\Release\App.g.cs:line 0 
" 
Sorry if its long, I dont exactly know how to shorten it :p 
I looked through the build.bat results and I figured out that I had to install WiX for it to properly install
Also, crimson ice blocks aren't dissapearing.

**Comment by Walkman100** [comment]
@rocket4wheel Please wrap that as a code block, like this:
 
    ```
    <code>
    ```
(Documentation [here](https://guides.github.com/features/mastering-markdown/#GitHub-flavored-markdown))

**Comment by enigmaquip** [comment]
That error of @rocket4wheel is throwing an error on it parsing the settings.xml..... Just looked through it and find no errors.....

**Comment by rocket4wheel** [comment]
@Walkman100  Thanks, didnt know how to.
Also, if its any info, That error was on speeders fork, i tried SBLogic's after and it worked well.

**Comment by Walkman100** [comment]
Okay, then if you got it working, maybe you should edit it out? (Make sure you used three backticks in a row, followed by a new line - it looks even worse right now than with no formatting :/ )

**Comment by IwantToknow** [comment]
When you will finish the work and release it?

**Comment by TheHoodedTeddy13** [comment]
Don't rush the magic computer people

**Comment by blacksteve13** [comment]
Guys please refrain from posting here asking for etas. Asking for etas is really rude to devs & contributors. So if you are reading this and havent signed for this site, and your only reason for coming here was to ask for an eta stop and dont sign up.Just skim threw here at the stuff you may or may not understand and wait til its done

**Comment by TheHoodedTeddy13** [comment]
Is there any way we less-knowledgable  people can help in the meantime?  Or is just leaving you in peace help enough?

**Comment by speeder** [comment]
Our biggest problem now is the missing original author, although I collected all the work done on my personal fork, and the editor mostly works (it can load and save files with some new 1.3 features, but actually editing those features, for example removing training dummies them breaks the save file) I don't want to make a non-official release, non-official releases tend to sometimes get accidentally promoted to be their own software (ie: sometimes they get more popular than the official version), and I doubt any of us contributors has enough time to take care of that, and non-official releases tend to swamp the original creator with bug complaints that are not his fault.

So we are waiting for the official dev to show up, merge our code and make an official release, so there is not much that can be done now (even from the programmers perspective).

**Comment by Walkman100** [comment]
> Our biggest problem now is the missing original author

When he does get back, he should add at least one more person to the TEdit Org (that's the point of Orgs, more than one owner)

**Comment by TheHoodedTeddy13** [comment]
Thanks for the information.  I use the program so much, I felt obligated to at least ask if I could help.  I do appreciate all you guys are doing.

**Pull request merged**